gtsfm/loader/hilti_loader.py

```diff
@@ -77,29 +88,133 @@ def __init__(
+
+        logger.info("[hilti_loader] Number of constraints: %d", len(self._constraints))
 
         # Read the poses for the IMU for rig indices from g2o file.
         self._w_T_imu: Dict[int, Pose3] = self.__read_lidar_pose_priors()
 
-        logger.info("Loading %d timestamps", self.num_rig_poses)
-        logger.info("Lidar camera available for %d timestamps", len(self._w_T_imu))
+        logger.info("[hilti_loader] Loading %d timestamps", self.num_rig_poses)
+        logger.info("[hilti_loader] Lidar camera available for %d timestamps", len(self._w_T_imu))
 
-    def __load_constraints(self) -> List[Constraint]:
+    def _load_constraints(self) -> List[Constraint]:
+        """Read all constraints but discard any out of the requested max length."""
+        # read constraints
         constraints_path = self._base_folder / LIDAR_CONSTRAINTS_RELATIVE_PATH
         constraints = Constraint.read(str(constraints_path))
-
         # filter them according to max length
-        constraints = list(filter(lambda c: c.a < self.num_rig_poses and c.b < self.num_rig_poses, constraints))
+        return [c for c in constraints if c.a < self.num_rig_poses and c.b < self.num_rig_poses]
+
+    @staticmethod
+    def _check_covariance(constraint: Constraint, overconfidence_threshold=1e-6) -> bool:
+        """Return false if covariance is bad."""
+        cov = constraint.cov
+        if np.isnan(cov).any() or np.min(np.diag(cov)) < overconfidence_threshold:
+            return False
+        try:
+            info = np.linalg.inv(cov)
+            if np.isnan(info).any():
+                return False
+            else:
+                return True
+        except np.linalg.LinAlgError:
+            return False
+
+    @classmethod
+    def _covariance_filtered_constraints(cls, constraints: List[Constraint]) -> Iterable[Constraint]:
+        """Filter constraints to remove those with invalid covariances."""
+        return filter(cls._check_covariance, constraints)
+
+    @classmethod
+    def _error_filtered_constraints(
+        cls, constraints: Iterable[Constraint], poses: List[Pose3], error_threshold=1e8
+    ) -> Iterable[Constraint]:
+        """Filter constraints to remove those with high errors (not used for now)."""
+
+        def acceptable_error(constraint):
+            a, b = constraint.a, constraint.b
+            predicted_aTb = poses[a].between(poses[b])
+            error = np.linalg.norm(constraint.aTb.logmap(predicted_aTb))
+            return error <= error_threshold
+
+        return filter(acceptable_error, constraints)
+
+    @staticmethod
+    def _update_stationary_constraints(constraints: List[Constraint]) -> List[Constraint]:
+        """Detect stationary periods and replace constraints there with 'hard' identity constraints."""
+        TRANSLATION_THRESHOLD = 0.02
+        ROTATION_THRESHOLD = np.deg2rad(0.5)
+
+        def stationary(c):
```

Comment:
type hints would be useful here. or maybe change function name to is_stationary()

Reply:
Added